README.md
# Electron Shutdown Command

Quickly shutdown, reboot, log off, halt, restarts, etc. your computer using the native shutdown command easily from an electron app. Cross platform (Windows; macOS and Linux)


## ATTENTION macOS and Linux Users:
**You need to run the app as sudo/root to allow to run the shutdown command!**


## Key Features
* **Secure and lightweight:** Uses only node/electron native modules

* **Written in TypeScript**

* **Well documented and easy to use**


## Installation
```
npm install --save electron-shutdown-command
```
or

```
yarn add electron-shutdown-command
```


## Usage

```
const shutdown = require('electron-shutdown-command');

shutdown.shutdown(); // simple system shutdown with default options
```
or
```
import * as shutdown from 'electron-shutdown-command';

shutdown.shutdown(); // simple system shutdown with default options
```
or

```
/* shutdown after 60 seconds, force all windows to close (Microsoft Windows only),
 * run as sudo (macOS and Linux only), debug command to console
 * (not execute it) and quit app after the command execution
 */
shutdown.shutdown({
  force: true,
  timerseconds: 60,
  sudo: true,
  debug: true,
  quitapp: true
})
```
or
```
// Warning: Microsoft Windows only
shutdown.logoff();
```


## Methods

### shutdown([options])
> shutdown / power-off your machine

### reboot([options])
> Reboot / restarts your machine

### logoff([options])
> Ends current session *Windows only*

### sleep([options])
> Enters sleep mode *macOS only*

### hibernate([options])
> Hibernate *Windows only*

### abort([options])
> Aborts or cancels a pending shutdown (this does not apply to "shutdown now", which does not wait before shutting down) *Windows and Linux only*

## Options

Property         | Type     | Default    | Description
---------------- | -------- | ---------- | ----------------------
`force`  | `boolean` | `false`     | Forces running applications to close *Windows only*
`sudo`  | `boolean` | `false`     | Run command as sudo *macOS and Linux*
`debug`  | `boolean` | `false`     | Shows shutdown command on console for debugging purposes **NOTE: It does not run it**
`quitapp`  | `boolean` | `false`     |  Quits your app after the shutdown command
`timerseconds`  | `number` | `0`     |  Sets the timer (value in seconds). **NOTES**: in *macOS* the minimum is 1 minute. The logoff, sleep and hibernate option cannot be scheduled, they are executed immediately


## Motivation and history
I quickly wrote this library because I need to add a shutdown option to an Electron Windows App. This is a very simple library that justs executes the shutdown command on different OS. **Warning:** I didn't have enough time to fully test this module on macOS and Linux.  If there is enough interest in some specific platform I can try to improve this module (I currently use this module on Windows 10).

## Contribution
Please send pull requests improving the usage and fixing bugs, improving documentation and providing better examples, or providing some tests, because these things are important.


## Learn More
Shutdown command
 * Windows: https://technet.microsoft.com/en-us/library/bb491003.aspx?f=255&MSPPError=-2147217396
 * macOS: https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man8/shutdown.8.html
 * Linux: https://www.computerhope.com/unix/ushutdow.htm

 * [Create Shortcuts on the Desktop to Run Programs as Root in Ubuntu 11.10](https://www.howtogeek.com/112700/create-shortcuts-on-the-desktop-to-run-programs-as-root-in-ubuntu-11.10/)
 * [How to Run GUI Apps as root in Mac OS X](https://osxdaily.com/2013/02/06/how-to-run-gui-apps-as-root-in-mac-os-x/)

## License
- Licensed under MIT

- Copyright (c) 2017-2020 [Samuel Carreira]

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [2.0.1] - 2020-10-06
### Changed
- Updated dependencies
- Add default parameter value (thanks @brunobg)

## [2.0.0] - 2020-07-20
### Changed
- Completely rewrite in TypeScript

## [0.0.7] - 2017-09-13
### Fixed
- Fixed abort bug (thanks @fabianbehrendt)

index.js
'use strict';

const cp = require('child_process');
const electron = require('electron');
const app = electron.app;

// learn more - documentation:
// windows: https://technet.microsoft.com/en-us/library/bb491003.aspx?f=255&MSPPError=-2147217396
// macos: https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man8/shutdown.8.html 
// linux: https://www.computerhope.com/unix/ushutdow.htm

/**
 * Shutdown command
 * 
 * @param {object} options 
 */
function shutdown(options = {}) {
    if (process.platform !== 'linux' &&
        process.platform !== 'darwin' &&
        process.platform !== 'win32') {
        throw new Error('Unknown or unsupported OS!');
    }

    let cmdarguments = ['shutdown'];

    if (process.platform === 'linux' || process.platform === 'darwin') {
        if (options.sudo) {
            cmdarguments.unshift('sudo');
        }

        cmdarguments.push('-h'); // halt
    }

    if (process.platform === 'win32') {
        cmdarguments.push('-s'); // shutdown

        if (options.force) {
            cmdarguments.push('-f');
        }      
    }

    cmdarguments.push(setTimer(options.timerseconds));

    executeCmd(cmdarguments, options.debug, options.quitapp);
}

/**
 * Hibernate (Windows only)
 * @param {Object} options 
 */
function hibernate(options = {}) {
    if (process.platform !== 'win32') {
        throw new Error('Unsupported OS (only Windows is supported)!');
    }

    let cmdarguments = ['shutdown'];

    cmdarguments.push('-h'); // hibernate    
 
    executeCmd(cmdarguments, options.debug, options.quitapp);
}

/**
 * Ends current session (Windows only)
 * @param {object} options 
 */
function logoff(options = {}) {
    if (process.platform !== 'win32') {
        throw new Error('Unsupported OS (only Windows is supported)!');
    }

    let cmdarguments = ['shutdown'];

    cmdarguments.push('-l'); // logoff    

    executeCmd(cmdarguments, options.debug, options.quitapp);
}

/**
 * Enters sleep mode (macOS Only)
 */
function sleep(options = {}) {
    if (process.platform !== 'darwin') {
        throw new Error('Unsupported OS (only macOS is supported)!');
    }

    let cmdarguments = ['shutdown'];
    
    cmdarguments.push('-s'); // sleep
    
    //cmdarguments.push(setTimer(options.timerseconds));

    executeCmd(cmdarguments, options.debug, options.quitapp);
}

/**
 * Aborts current scheduled shutdown
 * @param {Object} options 
 */
function abort(options = {}) {
    if (process.platform !== 'win32' && process.platform !== 'linux') {
        throw new Error('Unsupported OS (only Windows and Linux are supported)!');
    }

    let cmdarguments = ['shutdown'];
    
    if (process.platform === 'win32') {
        cmdarguments.push('-a'); // abort
    } else {
        // linux
        cmdarguments.push('-c'); // cancel a pending shutdown
    }
    
    executeCmd(cmdarguments, options.debug, options.quitapp);
}

/**
 * Shutdown / power-off your machine
 * @param {Object} options 
 */
function reboot(options = {}) {
    if (process.platform !== 'linux' &&
        process.platform !== 'darwin' &&
        process.platform !== 'win32') {
        throw new Error('Unknown or unsupported OS!');
    }

    let cmdarguments = ['shutdown'];

    cmdarguments.push('-r'); // reboot

    if (process.platform === 'linux' || process.platform === 'darwin') {
        if (options.sudo) {
            cmdarguments.unshift('sudo');
        }
    }

    if (process.platform === 'win32') {
        if (options.force) {
            cmdarguments.push('-f');
        }      
    }

    cmdarguments.push(setTimer(options.timerseconds));

    executeCmd(cmdarguments, options.debug, options.quitapp);
}

/**
 * set current time argument
 * @param {Number} timervalue value in seconds to delay, false to execute now
 */
function setTimer(timervalue) {
    let cmdarguments = [];
    
    if (process.platform === 'linux' ) {
        if (timervalue) {
            cmdarguments.push(timervalue.toString());
        } else {
            cmdarguments.push('now');
        }
    }

    if (process.platform === 'darwin') {
        if (timervalue) {
			let timeinminutes = Math.round(Number(timervalue) / 60);
			if (timeinminutes === 0) {
				timeinminutes = 1; // min 1 minute
			}
            cmdarguments.push('-t ' + timeinminutes.toString());
        } else {
            cmdarguments.push('now');
        }
    }

    if (process.platform === 'win32') {
        if (timervalue) {
            cmdarguments.push('-t ' + timervalue.toString());
        } else {
            cmdarguments.push('-t 0'); // set to 0 because default is 20 seconds
        }
    }

    return cmdarguments;
} 


/**
 * Execute command
 * @param {Object} cmd array of commands 
 * @param {Boolean} debug show console results, not execute the command
 * @param {Boolean} quitapp quit app after send the command
 */
function executeCmd(cmd, debug, quitapp) {
    if (typeof cmd !== 'object') {
        throw new Error('Empty or invalid argument list!');
    }

    let finalcmd = cmd.join(' ');

    if (debug) {
        // debug mode, print to console
        console.log(`Debug shutdown command: ${finalcmd}`)
        if (quitapp) {
            console.log("Now exit app...");
            app.exit(0);
        }
    } else {
        cp.exec(finalcmd, (err, stdout, stderr) => {
            if (err) {
                console.error(err);
                return;
            }
            //console.log(stdout);
            if (quitapp) {
                app.exit(0);
            }
        });
    }
}

module.exports = {
    shutdown: shutdown,
    reboot: reboot,
    hibernate: hibernate,
    logoff: logoff,
    sleep: sleep,
    abort: abort
};

src/index.ts
/*!
 * Electron Shutdown Command
 *
 * Learn more - documentation:
 * windows: https://technet.microsoft.com/en-us/library/bb491003.aspx?f=255&MSPPError=-2147217396
 * macos: https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man8/shutdown.8.html
 * linux: https://www.computerhope.com/unix/ushutdow.htm
 *
 * Licensed under MIT
 * Copyright (c) 2020 [Samuel Carreira]
 */

import cp = require('child_process');
import {app} from 'electron';

export interface ElectronShutdownCommandUserOptions {
  /**
   * Sets the timer (value in seconds).
   * Notes: in macOS the minimum is 1 minute. The logoff, sleep
   * and hibernate option cannot be scheduled, it executes immediately
   *
   * @default 0
   */
  timerseconds?: number;
  /**
   * Quits your app after the shutdown command
   *
   * @default false
   */
  quitapp?: boolean;
  /**
   * Run as sudo *macOS and Linux only*
   *
   * @default false
   */
  sudo?: boolean;
  /**
   * Forces running applications to close *Windows only*
   *
   * @default false
   */
  force?: boolean;
  /**
   * Shows shutdown command on console for debugging
   * purposes **NOTE: It does not run it**
   *
   * @default false
   */
  debug?: boolean;
}

interface ElectronShutdownCommandOptions {
  timerseconds: number;
  quitapp: boolean;
  sudo: boolean;
  force: boolean;
  debug: boolean;
}

function applyDefaults(options: ElectronShutdownCommandUserOptions): ElectronShutdownCommandOptions {
  const defaultOptions: ElectronShutdownCommandOptions = {
    timerseconds: 0,
    quitapp: false,
    sudo: false,
    force: false,
    debug: false
  };

  return {...defaultOptions, ...options};
}

/**
 * Shutdown command
 *
 * @param {object} options
 */
export function shutdown(options: ElectronShutdownCommandUserOptions = {}) {
  const validatedOptions: ElectronShutdownCommandOptions = applyDefaults(options);

  const cmdarguments = ['shutdown'];

  if (process.platform === 'linux' || process.platform === 'darwin') {
    if (options.sudo) {
      cmdarguments.unshift('sudo');
    }

    cmdarguments.push('-h'); // Halt
  }

  if (process.platform === 'win32') {
    cmdarguments.push('-s'); // Shutdown

    if (options.force) {
      cmdarguments.push('-f');
    }
  }

  cmdarguments.push(...setTimer(validatedOptions.timerseconds));

  executeCmd(cmdarguments, validatedOptions.debug, validatedOptions.quitapp);
}

/**
 * Hibernate (Windows only)
 * @param {Object} options
 */
export function hibernate(options: ElectronShutdownCommandUserOptions = {}) {
  if (process.platform !== 'win32') {
    throw new Error('Unsupported OS (only Windows is supported)!');
  }

  const validatedOptions: ElectronShutdownCommandOptions = applyDefaults(options);

  const cmdarguments = ['shutdown'];

  cmdarguments.push('-h'); // Hibernate

  executeCmd(cmdarguments, validatedOptions.debug, validatedOptions.quitapp);
}

/**
 * Ends current session (Windows only)
 * @param {object} options
 */
export function logoff(options: ElectronShutdownCommandUserOptions = {}) {
  if (process.platform !== 'win32') {
    throw new Error('Unsupported OS (only Windows is supported)!');
  }

  const validatedOptions: ElectronShutdownCommandOptions = applyDefaults(options);

  const cmdarguments = ['shutdown'];

  cmdarguments.push('-l'); // Logoff

  executeCmd(cmdarguments, validatedOptions.debug, validatedOptions.quitapp);
}

/**
 * Enters sleep mode (macOS Only)
 */
export function sleep(options: ElectronShutdownCommandUserOptions = {}) {
  if (process.platform !== 'darwin') {
    throw new Error('Unsupported OS (only macOS is supported)!');
  }

  const validatedOptions: ElectronShutdownCommandOptions = applyDefaults(options);

  const cmdarguments = ['shutdown'];

  cmdarguments.push('-s'); // Sleep

  // cmdarguments.push(setTimer(options.timerseconds));

  executeCmd(cmdarguments, validatedOptions.debug, validatedOptions.quitapp);
}

/**
 * Aborts current scheduled shutdown
 * @param {Object} options
 */
export function abort(options: ElectronShutdownCommandUserOptions = {}) {
  if (process.platform !== 'win32' && process.platform !== 'linux') {
    throw new Error('Unsupported OS (only Windows and Linux are supported)!');
  }

  const validatedOptions: ElectronShutdownCommandOptions = applyDefaults(options);

  const cmdarguments = ['shutdown'];

  if (process.platform === 'win32') {
    cmdarguments.push('-a'); // Abort
  } else {
    // Linux
    cmdarguments.push('-c'); // Cancel a pending shutdown
  }

  executeCmd(cmdarguments, validatedOptions.debug, validatedOptions.quitapp);
}

/**
 * Shutdown / power-off your machine
 * @param {Object} options
 */
export function reboot(options: ElectronShutdownCommandUserOptions = {}) {
  const validatedOptions: ElectronShutdownCommandOptions = applyDefaults(options);

  const cmdarguments = ['shutdown'];

  cmdarguments.push('-r'); // Reboot

  if (process.platform === 'linux' || process.platform === 'darwin') {
    if (options.sudo) {
      cmdarguments.unshift('sudo');
    }
  }

  if (process.platform === 'win32') {
    if (options.force) {
      cmdarguments.push('-f');
    }
  }

  cmdarguments.push(...setTimer(validatedOptions.timerseconds));

  executeCmd(cmdarguments, validatedOptions.debug, validatedOptions.quitapp);
}

/**
 * Set current time argument
 * @param {Number} timervalue value in seconds to delay, false to execute now
 */
function setTimer(timervalue: number): string[] {
  const cmdarguments = [];

  if (process.platform === 'linux') {
    if (timervalue) {
      cmdarguments.push(timervalue.toString());
    } else {
      cmdarguments.push('now');
    }
  }

  if (process.platform === 'darwin') {
    if (timervalue) {
      let timeinminutes = Math.round(Number(timervalue) / 60);
      if (timeinminutes === 0) {
        timeinminutes = 1; // Minimum 1 minute
      }

      cmdarguments.push(`-t ${timeinminutes.toString()}`);
    } else {
      cmdarguments.push('now');
    }
  }

  if (process.platform === 'win32') {
    if (timervalue) {
      cmdarguments.push(`-t ${timervalue.toString()}`);
    } else {
      cmdarguments.push('-t 0'); // Set to 0 because default is 20 seconds
    }
  }

  return cmdarguments;
}

/**
 * Execute command

 * @param {Boolean} debug show console results, not execute the command
 * @param {Boolean} quitapp quit app after send the command
 */
function executeCmd(cmd: string[], debug: boolean, quitapp: boolean): void {
  const finalcmd = cmd.join(' ');

  if (debug) {
    // Debug mode, print to console
    console.log(`Debug shutdown command: ${finalcmd}`);
    if (quitapp) {
      console.log('Now exit app...');
      app.exit(0);
    }

    return;
  }

  // Execute the command
  cp.exec(finalcmd, err => {
    if (err) {
      console.error(err);
      return;
    }

    if (quitapp) {
      app.exit(0);
    }
  });
}
